jsonschema/src/schema_node.rs

```diff
@@ -0,0 +1,429 @@
+use crate::{
+    compilation::context::CompilationContext,
+    error::ErrorIterator,
+    keywords::BoxedValidator,
+    output::{Annotations, BasicOutput, ErrorDescription, OutputUnit},
+    paths::{AbsolutePath, InstancePath, JSONPointer},
+    validator::{format_validators, PartialApplication, Validate},
+    JSONSchema,
+};
+use ahash::AHashMap;
+use std::fmt;
+
+/// A node in the schema tree, returned by [`compile_validators`]
+#[derive(Debug)]
+pub(crate) struct SchemaNode {
+    validators: NodeValidators,
+    relative_path: JSONPointer,
```

Comment:
In many validators, there are `schema_path` fields (e.g. `OneOfValidator`). Do we need them when `SchemaNode` is also present? Mostly thinking in terms of whether the path we need for `ValidationError` could be taken from here

Reply:
This definitely could be done. My instinct is that the best way to actually do it is to start trying to reimplement `validate` using `apply` for applicator keywords. That reduces some of the other logic duplication in applicator keywords and provides a nice avenue to start experimenting with an arena allocator for error collection. i.e I think we should leave it out of this PR and address it shortly in another PR.

I'm happy to address in this PR if you feel it shouldn't be left until later though.

Reply:
Cool! I think it could be done separately :)